File: ports/kiit-codes-ts/test/formats/problem-converter.test.ts
import { describe, expect, it } from "vitest";
import { ProblemConverter, problemFor } from "../../src/formats/problem-converter.js";
import { ErrorDetail } from "../../src/formats/error-item.js";
import type { ErrorItem } from "../../src/formats/error-item.js";
import { CodesToHttp } from "../../src/codes.js";
import { ErrorField, ErrorList } from "../../src/err.js";
import { Rejected, Restricted, Invalid, StatusConstants } from "../../src/groups.js";

// Ported from formats/ProblemConverterTest.kt.

describe("ProblemConverter", () => {
  const converter = ProblemConverter();

  it("baseUrl defaults for the KIIT origin, with code as a query param", () => {
    const status = Restricted("PAYMENT_REQUIRES_3DS", "3DS required", StatusConstants.KIIT);
    expect(converter.convert(status).type).toBe(
      "https://www.kiit.dev/docs/kiit-codes?code=Failed:Restricted:PAYMENT_REQUIRES_3DS#taxonomy",
    );
  });

  it("a domain origin with no entry builds type from the origin", () => {
    const status = Restricted("PAYMENT_REQUIRES_3DS", "3DS required", "stripe.com", "payments.cards");
    expect(converter.convert(status).type).toBe(
      "https://stripe.com/problems/payments.cards/restricted/payment-requires-3ds",
    );
  });

  it("a plain id origin with no entry does not throw and is used as is", () => {
    const status = Rejected("OUT_OF_STOCK", "Out of stock", "myapp1");
    expect(converter.convert(status).type).toBe("https://myapp1/problems/rejected/out-of-stock");
  });

  it("the origin is lowercased in the derived baseUrl", () => {
    const status = Rejected("OUT_OF_STOCK", "Out of stock", "MyApp1");
    expect(converter.convert(status).type).toBe("https://myapp1/problems/rejected/out-of-stock");
  });

  it("an empty scope is skipped and a set scope is the first segment", () => {
    const noScope = Rejected("OUT_OF_STOCK", "Out of stock", "stripe.com");
    const scoped = { ...noScope, scope: "payments.cards" };
    expect(converter.convert(noScope).type).toBe("https://stripe.com/problems/rejected/out-of-stock");
    expect(converter.convert(scoped).type).toBe("https://stripe.com/problems/payments.cards/rejected/out-of-stock");
  });

  it("convert uses the registered baseUrl and the full type path", () => {
    const problems = ProblemConverter({ "stripe.com": "https://stripe.com/errors" });
    const status = Restricted("PAYMENT_REQUIRES_3DS", "3DS required", "stripe.com", "payments.cards");
    expect(problems.convert(status).type).toBe(
      "https://stripe.com/errors/payments.cards/restricted/payment-requires-3ds",
    );
  });

  it("a registered entry wins over the origin-derived baseUrl", () => {
    const problems = ProblemConverter({ "stripe.com": "https://docs.stripe.com/errors" });
    const status = Rejected("DUPLICATE_CHARGE", "Already processed", "stripe.com", "payments.cards");
    expect(problems.convert(status).type).toBe(
      "https://docs.stripe.com/errors/payments.cards/rejected/duplicate-charge",
    );
  });

  it("registered keys are lowercased and a trailing slash is trimmed", () => {
    const problems = ProblemConverter({ "Stripe.com": "https://docs.stripe.com/errors/" });
    const status = Rejected("DUPLICATE_CHARGE", "Already processed", "stripe.com");
    expect(problems.convert(status).type).toBe("https://docs.stripe.com/errors/rejected/duplicate-charge");
  });

  it("the KIIT origin is fixed and cannot be overridden", () => {
    const problems = ProblemConverter({ [StatusConstants.KIIT]: "https://example.com/problems" });
    expect(problems.convert(Restricted.DENIED).type).toBe(
      "https://www.kiit.dev/docs/kiit-codes?code=Failed:Restricted:DENIED#taxonomy",
    );
  });

  it("a custom typeBuilder suffix is appended to the origin-derived baseUrl", () => {
    const status = Rejected("DUPLICATE_CHARGE", "Already processed", "stripe.com");
    const problem = converter.convert(status, undefined, (s) => `errors/${s.name.toLowerCase()}`);
    expect(problem.type).toBe("https://stripe.com/problems/errors/duplicate_charge");
  });

  it("accepts an explicit mapping with no baseUrls", () => {
    expect(ProblemConverter({}, CodesToHttp()).convert(Restricted.FORBIDDEN).status).toBe(403);
  });

  it("convertWithUrl uses the explicit baseUrl regardless of baseUrls", () => {
    const status = Restricted("PAYMENT_REQUIRES_3DS", "3DS required", "stripe.com");
    const problem = converter.convertWithUrl(status, undefined, "https://example.com/probs");
    expect(problem.type).toBe("https://example.com/probs/restricted/payment-requires-3ds");
  });

  it("convertWithUrl trims a trailing slash", () => {
    const status = Restricted("PAYMENT_REQUIRES_3DS", "3DS required", "stripe.com");
    const problem = converter.convertWithUrl(status, undefined, "https://example.com/probs/");
    expect(problem.type).toBe("https://example.com/probs/restricted/payment-requires-3ds");
  });

  it("convertCustomWithUrl maps errors and uses the explicit baseUrl", () => {
    const status = Invalid("BAD_EMAIL", "Bad email", "stripe.com");
    const err = ErrorList([ErrorField("email", "x", "Invalid email")], "Validation failed");
    const problem = converter.convertCustomWithUrl(status, err, "https://example.com/probs", (e) =>
      ErrorDetail("mapped", e.message),
    );
    expect(problem.type).toBe("https://example.com/probs/invalid/bad-email");
    expect(problem.errors?.[0]).toEqual(ErrorDetail("mapped", "Invalid email"));
  });

  it("title is the status message, status is the HTTP code", () => {
    const problem = converter.convert(Restricted.FORBIDDEN);
    expect(problem.title).toBe(Restricted.FORBIDDEN.message);
    expect(problem.status).toBe(403);
  });

  it("no err leaves detail/instance/errors undefined", () => {
    const problem = converter.convert(Restricted.DENIED);
    expect(problem.detail).toBeUndefined();
    expect(problem.instance).toBeUndefined();
    expect(problem.errors).toBeUndefined();
  });

  it("an ErrorList populates default ErrorDetail entries", () => {
    const err = ErrorList([ErrorField("email", "not-an-email", "Invalid email")], "Validation failed");
    const problem = converter.convert(Invalid.INVALID_VALUE, err);
    expect(problem.detail).toBe("Validation failed");
    expect(problem.errors?.[0]).toEqual(ErrorDetail("email", "Invalid email"));
  });

  it("convertCustom maps through a custom ErrorItem", () => {
    interface RichError extends ErrorItem {
      readonly field?: string;
      readonly message: string;
      readonly ref?: unknown;
    }
    const err = ErrorList([ErrorField("email", "not-an-email", "Invalid email", undefined, "req-42")], "Validation failed");
    const problem = converter.convertCustom<RichError>(Invalid.INVALID_VALUE, err, (e) => ({
      field: e.kind === "ErrorField" ? e.field : undefined,
      message: e.message,
      ref: e.ref,
    }));
    expect(problem.errors?.[0]?.ref).toBe("req-42");
  });

  it("problemFor matches convert with the default ErrorDetail", () => {
    const status = Restricted.FORBIDDEN;
    expect(problemFor(status)).toEqual(converter.convert(status));
  });
});

File: kiit-codes-kotlin/kiit-codes/src/commonTest/kotlin/kiit/codes/formats/ProblemConverterTest.kt
package kiit.codes.formats

import kiit.codes.CodesToHttp
import kiit.codes.Err
import kiit.codes.Failed
import kiit.codes.Invalid
import kiit.codes.Restricted
import kiit.codes.StatusConstants
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class ProblemConverterTest {
    private val converter = ProblemConverter()

    @Test
    fun baseUrlDefaultsForKiitOriginWithCodeAsQueryParam() {
        val status = Failed.Restricted("PAYMENT_REQUIRES_3DS", "3DS required", origin = StatusConstants.KIIT)
        val problem = converter.convert(status)
        assertEquals(
            "https://www.kiit.dev/docs/kiit-codes?code=Failed:Restricted:PAYMENT_REQUIRES_3DS#taxonomy",
            problem.type,
        )
    }

    @Test
    fun domainOriginWithNoEntryBuildsTypeFromOrigin() {
        val status = Failed.Restricted("PAYMENT_REQUIRES_3DS", "3DS required", origin = "stripe.com", scope = "payments.cards")
        assertEquals(
            "https://stripe.com/problems/payments.cards/restricted/payment-requires-3ds",
            converter.convert(status).type,
        )
    }

    @Test
    fun plainIdOriginWithNoEntryDoesNotThrowAndIsUsedAsIs() {
        val status = Failed.Rejected("OUT_OF_STOCK", "Out of stock", origin = "myapp1")
        assertEquals("https://myapp1/problems/rejected/out-of-stock", converter.convert(status).type)
    }

    @Test
    fun originIsLowercasedInTheDerivedBaseUrl() {
        val status = Failed.Rejected("OUT_OF_STOCK", "Out of stock", origin = "MyApp1")
        assertEquals("https://myapp1/problems/rejected/out-of-stock", converter.convert(status).type)
    }

    @Test
    fun emptyScopeIsSkippedAndScopeIsTheFirstSegmentWhenSet() {
        val noScope = Failed.Rejected("OUT_OF_STOCK", "Out of stock", origin = "stripe.com")
        val scoped = noScope.copy(scope = "payments.cards")
        assertEquals("https://stripe.com/problems/rejected/out-of-stock", converter.convert(noScope).type)
        assertEquals("https://stripe.com/problems/payments.cards/rejected/out-of-stock", converter.convert(scoped).type)
    }

    @Test
    fun convertUsesRegisteredBaseUrlAndFullTypePath() {
        val problems = ProblemConverter(mapOf("com.stripe" to "https://stripe.com/problems"))
        val status = Failed.Restricted("PAYMENT_REQUIRES_3DS", "3DS required", origin = "com.stripe", scope = "payments.cards")
        assertEquals("https://stripe.com/problems/payments.cards/restricted/payment-requires-3ds", problems.convert(status).type)
    }

    @Test
    fun registeredEntryWinsOverTheOriginDerivedBaseUrl() {
        val problems = ProblemConverter(mapOf("stripe.com" to "https://docs.stripe.com/errors"))
        val status = Failed.Rejected("DUPLICATE_CHARGE", "Already processed", origin = "stripe.com", scope = "payments.cards")
        assertEquals("https://docs.stripe.com/errors/payments.cards/rejected/duplicate-charge", problems.convert(status).type)
    }

    @Test
    fun registeredKeysAreLowercasedAndTrailingSlashIsTrimmed() {
        val problems = ProblemConverter(mapOf("Stripe.com" to "https://docs.stripe.com/errors/"))
        val status = Failed.Rejected("DUPLICATE_CHARGE", "Already processed", origin = "stripe.com")
        assertEquals("https://docs.stripe.com/errors/rejected/duplicate-charge", problems.convert(status).type)
    }

    @Test
    fun kiitOriginIsFixedAndCannotBeOverridden() {
        val problems = ProblemConverter(mapOf(StatusConstants.KIIT to "https://example.com/problems"))
        assertEquals(
            "https://www.kiit.dev/docs/kiit-codes?code=Failed:Restricted:DENIED#taxonomy",
            problems.convert(Restricted.DENIED).type,
        )
    }

    @Test
    fun customTypeBuilderSuffixIsAppendedToTheOriginDerivedBaseUrl() {
        val status = Failed.Rejected("DUPLICATE_CHARGE", "Already processed", origin = "stripe.com")
        val problem = converter.convert(status, typeBuilder = { "errors/${it.name.lowercase()}" })
        assertEquals("https://stripe.com/problems/errors/duplicate_charge", problem.type)
    }

    @Test
    fun acceptsAnExplicitMappingWithNoBaseUrls() {
        val problems = ProblemConverter(mapping = CodesToHttp())
        assertEquals(403, problems.convert(Restricted.FORBIDDEN).status)
    }

    @Test
    fun convertWithUrlTrimsATrailingSlash() {
        val status = Failed.Restricted("PAYMENT_REQUIRES_3DS", "3DS required", origin = "com.stripe")
        val problem = converter.convertWithUrl(status, baseUrl = "https://example.com/probs/")
        assertEquals("https://example.com/probs/restricted/payment-requires-3ds", problem.type)
    }

    @Test
    fun convertCustomWithUrlMapsErrorsAndUsesTheExplicitBaseUrl() {
        val status = Failed.Invalid("BAD_EMAIL", "Bad email", origin = "stripe.com")
        val err = Err.ErrorList(errors = listOf(Err.ErrorField("email", "x", "Invalid email")), message = "Validation failed")
        val problem = converter.convertCustomWithUrl(status, err, "https://example.com/probs") { ErrorDetail("mapped", it.message) }
        assertEquals("https://example.com/probs/invalid/bad-email", problem.type)
        assertEquals(ErrorDetail("mapped", "Invalid email"), problem.errors?.single())
    }

    @Test
    fun titleIsStatusMessageAndStatusIsHttpCode() {
        val problem = converter.convert(Restricted.FORBIDDEN)
        assertEquals(Restricted.FORBIDDEN.message, problem.title)
        assertEquals(403, problem.status)
    }

    @Test
    fun noErrLeavesDetailAndInstanceAndErrorsNull() {
        val problem = converter.convert(Restricted.DENIED)
        assertNull(problem.detail)
        assertNull(problem.instance)
        assertNull(problem.errors)
    }

    @Test
    fun errorListPopulatesDefaultErrorDetailEntries() {
        val err =
            Err.ErrorList(
                errors = listOf(Err.ErrorField("email", "not-an-email", "Invalid email")),
                message = "Validation failed",
            )
        val problem = converter.convert(Invalid.INVALID_VALUE, err)
        assertEquals("Validation failed", problem.detail)
        assertEquals(ErrorDetail("email", "Invalid email"), problem.errors?.single())
    }

    @Test
    fun genericConvertMapsThroughACustomErrorItem() {
        data class RichError(override val field: String?, override val message: String, val ref: Any?) : ErrorItem

        val err =
            Err.ErrorList(
                errors = listOf(Err.ErrorField("email", "not-an-email", "Invalid email", ref = "req-42")),
                message = "Validation failed",
            )
        val problem =
            converter.convertCustom(Invalid.INVALID_VALUE, err) { RichError((it as? Err.ErrorField)?.field, it.message, it.ref) }
        assertEquals("req-42", (problem.errors?.single() as RichError).ref)
    }
}
